### PR TITLE
CP-23829: Preliminary TLS functionality

### PR DESCRIPTION
**Pull request by thomassa**
It works but some code is copied from the nbd repository,
rather than being shared.

### CODE REVIEW

**Comment by thomassa** [comment]
Depends on xapi-project/nbd#79 which adds TLS functionality to the nbd package.

**Comment by mseri** [comment]
The build is failing with
```
- File "src/main.ml", line 126, characters 2-27:
- Error: Unbound value Nbd_lwt_unix.with_channel
```

**Comment by thomassa** [comment]
Also with this change, xapi-nbd requires at least one command-line argument, so in the `xenserver-specs` repository's `xapi-nbd.service` file for systemd, the `ExecStart=/usr/sbin/xapi-nbd` start-up command will need to cater for this.

**Comment by mseri** [comment]
Can we make the parameter optional, falling back to the old behaviour?

**Comment by thomassa** [comment]
@mseri wrote
> The build is failing with `Error: Unbound value Nbd_lwt_unix.with_channel`

Yes, that's expected because it depends on the change in the nbd package.

**Comment by thomassa** [comment]
I'm reluctant to fall back to the old behaviour because that would be not using TLS.
(I suppose we could implement the fallback temporarily, so that we can change repositories one at a time rather than needing to synchronise the changes.)

**Comment by mseri** [comment]
Let's review and merge nbd first and then this one, so that we don't need the fallback

**Comment by thomassa** [comment]
I think it's best to review this and the `nbd` pull-request together so that the review includes the new/changed parts of the interface between them.

File: src/main.ml [hunk 1]
```diff
@@ -115,9 +123,8 @@ let handle_connection xen_api_uri fd =
       )
   in
 
-  let channel = Nbd_lwt_unix.of_fd fd in
-  Lwt.finalize
-    (fun () ->
+  Nbd_lwt_unix.with_channel fd tls_role
+    (fun channel ->
        Nbd_lwt_unix.Server.connect channel ()
```
Comment (gaborigloi) [inline src/main.ml]: Could we use `Server.with_connection` here? (First it needs to pass the export name to its argument function.)

**Review by gaborigloi — APPROVED** [review]
Looks good, just one thing: we should use the `Server.with_connection` function here.
I've realized that we can make the CLI more orthogonal and remove the require* and initi_tl_get_... functions by splitting it into serve-tls and serve-notls commands. I'll do that in a followup PR after this.

**Comment by mseri** [comment]
Please add the fix for travis. It should be enough to use `4.04.2` as the compiler

**Comment by thomassa** [comment]
I've added a second commit to use `Server.with_connection` though really that's unrelated to the addition of TLS functionality.
Now to change the compiler to make Travis happy...

**Comment by mseri** [comment]
I think with the script we use here it is version 4.04 :-/

**Comment by thomassa** [comment]
Trying with 4.04 now...
Hah. I see it's busy installing 4.04.2 anyway.

**Comment by mseri** [comment]
It's working. It is now failing only because we have not released the new nbd

**Comment by thomassa** [comment]
@mseri ...and that new release is defined in https://github.com/xapi-project/xs-opam/pull/134

**Comment by mseri** [comment]
Yes, I saw that. I think this can be merged.

**Pull request merged**